### PR TITLE
rfc: use PRs instead of issues for RFCs

### PR DESCRIPTION
**Pull request by admbtlr**
Proposing a new approach for RFCs.

### CODE REVIEW

File: RFCs/use-prs-instead-of-issues.md [hunk 1]
```diff
@@ -0,0 +1,8 @@
+# Use PRs instead of issues for RFCs
+
+We already have a [playbook](../playbooks/rfcs.md) for RFCs. We used to use GitHub [issues](https://github.com/artsy/README/issues?q=is%3Aissue+is%3Aclosed), but this approach has two drawbacks: 
+
+- it's impossible to add inline or threaded comments to an issue, and the conversation can become very hard to follow
```
Comment (pvinis) [inline RFCs/use-prs-instead-of-issues.md:5]: using this worked

> wow

but yea, that totally makes sense, cause now we can have many threads running!

**Comment by admbtlr** [comment]
> One thing i would like to have defined/cleared up, is if we should use the PR desc or just keep everything on the PR files. I think PR files makes more sense, but I would like it if all of us follow one way.
> 
> Another thing, is about how we handle changes/reviews. In my mind, it makes sense to **only** leave comments on the PR and comments on the PR files/lines. It seems strange to use approvals/request changes. Seems like the author should be the one responsible to either merge or close, with the final comment of the next steps. Is that what you have in mind? How are you thinking of all that?

The outcome of an RFC should be the new markdown document that the PR is proposing to add. This document represents the content of the RFC and, once merged, is itself a permanent record of the org's decision to accept and act on the RFC. 

So yes, the discussion should take place within the context of the document, and the PR description can be left empty. Or just include a brief summary of the document, as I did here.

And yes, it's probably easier to use 👍 / 👎 reactions on the PR description, rather than approving / requesting changes. And yes, it should be the author who merges or closes the PR, based on the votes and discussion.

**Comment by patrinoua** [comment]
love this 🔥

**Comment by damassi** [comment]
It just makes an enormous amount of sense 👍 Nice ideal @admbtlr.

**Comment by jacobherrington** [comment]
> The outcome of an RFC should be the new markdown document that the PR is proposing to add. This document represents the content of the RFC and, once merged, is itself a permanent record of the org's decision to accept and act on the RFC.

100% agree with this line of thinking.

**Comment by joeyAghion** [comment]
I'm very in favor of RFCs as PRs (and we've [had a few](https://github.com/artsy/README/pulls?q=is%3Apr+rfc+is%3Aclosed+label%3ARFC)). My preference is for the outcome to not just be a new doc in the `RFCs/` directory, but corresponding updates to our playbooks, tech radar, etc. _That's_ where the meat of significant process changes can be debated, and where a developer will look for succinct documentation rather than the whole of our RFC history. (Of course they can all be linked together, too.)

**Comment by gkartalis** [comment]
Love this idea, it makes so much sense! The best feature of this approach is the ability to let people provide in-context suggestions or changes! Thanks for bringing this up!

File: RFCs/use-prs-instead-of-issues.md [hunk 2]
```diff
@@ -0,0 +1,8 @@
+# Use PRs instead of issues for RFCs
```
[the remark this answers is not in the document]
Comment (gkartalis) [reply RFCs/use-prs-instead-of-issues.md:1]: > They're a whole new medium that we're not currently using, so there would be some cognitive overhead involved in introducing them (Having said that, maybe we should be using them more widely? Are there other needs we have that could be met with discussion? I think there's too much overlap with our Slack usage, and that discussions are best for wider communities rather than internal company discussions, but could be worthwhile investigating)

100% agree with this we don't need to put congitive overhead to anyone 👍 but worth investigating maybe at some point in the future? 🤔 

> You can't do inline, line-level comments - which is the main motivation for changing our approach in the first place

Well, we might be able to do it! Not with the way that we can do it on PRs but rather with the quote reply format ^ (pasting the line/code that we want to comment on) and the good thing with discussions is that we can start a thread from the comment like [this one](https://github.com/vercel/next.js/discussions/16832#discussioncomment-61499) that could act as an inline comment. 

Don't consider [this comment](https://github.com/artsy/README/pull/467#discussion_r899107249) blocking for this RFC, just wanted to put this into our radars because it looks super interesting!

**Comment by admbtlr** [comment]
> (@admbtlr seems like we should update the Markedown document this RFC introduces to include that part, right? 😉).

I added a little more detail in a new commit - does that cover it @mdole ?

**Comment by mdole** [comment]
LGTM! Thanks @admbtlr

File: playbooks/rfcs.md [hunk 3]
```diff
@@ -20,12 +20,12 @@ process with as much bang for buck as possible given our size.
 
 ### TLDR: Our process looks like:
 
-- Someone opens an issue on a repo that begins with "RFC: ", and that causes a message to be posted to Slack's #dev
-  room via Peril.
+- Someone opens a pull request on a repo (or an issue, if it's a repo that doesn't have an rfc directory to hold RFC docs) that begins with "RFC: "; this causes a message to be posted to Slack's #dev
```
Comment (erikdstock) [inline playbooks/rfcs.md:23]: I'm a little confused here so want to check my understanding:
- We can create RFCs from any repository, defaulting to an `/rfc` directory
- The rfc directory on the main branch is effectively a list of passed RFCs

Should engineers therefore view the rfcs directory as a directory of policies and decisions we've made, or should a given PR include changes elsewhere as well?

Comment (admbtlr) [reply playbooks/rfcs.md:23]: Your understanding is correct, but as far as I understand it, you're asking whether an RFC PR should also include any actions that may be taken to resolve the PR. Is that right?

So in this PR, I've included actions that are required to implement the RFC, i.e.
- create the `/rfc` dir
- update the playbook

This makes sense to me in this case because everything affected by this RFC is in the same repo. So the PR contains:
1. the case I made (the RFC itself)
2. the decision made by the org to accept the RFC (the comments)
3. the changes required to implement the RFC (the two actions listed above)

I could definitely see a case for separating (3) out into a separate PR, and I'm happy to do that if you think it makes more sense. OTOH it feels nicely atomic to have it all bundled in one place 🤷 

What do you think @erikdstock ? Which approach is preferable?

Comment (admbtlr) [reply playbooks/rfcs.md:23]: To clarify: by "which approach" I mean:

- include the changes required to implement the RFC inside the PR, along with the RFC itself, or
- restrict the PR to only include the RFC, create a new PR for the changes required to implement it

**Comment by pvinis** [comment]
should we have a `_template.md` file in that `RFCs` folder, for easy copy/pasting and editing?

File: playbooks/rfcs.md (hunk 3 above)
Comment (admbtlr) [reply playbooks/rfcs.md:23]: @erikdstock any thoughts?

Comment (erikdstock) [reply playbooks/rfcs.md:23]: I think my confusion was around:
a) now RFCs can happen on any repo? _Yes, and peril looks for PRs to a document in the `/rfc` folder or with a title starting with `RFC: `_
b) whether an RFC is accepted or not, what happens to the PR? _it should be merged at resolution with the merge documenting the result_

If these are correct that addresses it for me - I think it would be nice to have some markdown frontmatter showing whether an RFC passed or not.

Comment (admbtlr) [reply playbooks/rfcs.md:23]: ping @erikdstock, when you have a moment

Comment (erikdstock) [reply playbooks/rfcs.md:23]: ✅ Looks good, thanks!
